xds/src/main/java/io/grpc/xds/LocalityStore.java

```diff
@@ -497,5 +443,164 @@ protected Helper delegate() {
         return orcaReportingHelperWrapper.asHelper();
       }
     }
+
+    private final class PriorityManager {
+
+      private final List<List<XdsLocality>> priorityTable = new ArrayList<>();
+      private Map<XdsLocality, LocalityInfo> localityInfoMap = ImmutableMap.of();
+      private int currentPriority = -1;
+      private ScheduledHandle failOverTimer;
+
+      /**
+       * Updates the priority ordering of localities with the given collection of localities.
+       * Recomputes the current ready localities to be used.
+       */
+      void updateLocalities(Map<XdsLocality, LocalityInfo> localityInfoMap) {
+        this.localityInfoMap = localityInfoMap;
+        priorityTable.clear();
+        for (XdsLocality newLocality : localityInfoMap.keySet()) {
+          int priority = localityInfoMap.get(newLocality).priority;
+          while (priorityTable.size() <= priority) {
+            priorityTable.add(new ArrayList<XdsLocality>());
+          }
+          priorityTable.get(priority).add(newLocality);
+        }
+
+        currentPriority = -1;
+        failOver();
+      }
+
+      /**
+       * Refreshes the group of localities with the given priority. Recomputes the current ready
+       * localities to be used.
+       */
+      void updatePriorityState(int priority) {
+        if (priority == -1 || priority > currentPriority) {
+          return;
+        }
+        List<WeightedChildPicker> childPickers = new ArrayList<>();
+
+        ConnectivityState overallState = null;
+        for (XdsLocality l : priorityTable.get(priority)) {
+          if (!localityMap.containsKey(l)) {
+            initLocality(l);
+          }
+          LocalityLbInfo localityLbInfo = localityMap.get(l);
+          localityLbInfo.reactivate();
+          ConnectivityState childState = localityLbInfo.childHelper.currentChildState;
+          SubchannelPicker childPicker = localityLbInfo.childHelper.currentChildPicker;
+
+          overallState = aggregateState(overallState, childState);
+
+          if (READY == childState) {
+            childPickers.add(
+                new WeightedChildPicker(localityInfoMap.get(l).localityWeight, childPicker));
+          }
+        }
+
+        if (priority == currentPriority) {
+          updatePicker(overallState, childPickers);
+          if (overallState == READY) {
+            cancelFailOverTimer();
+          } else if (overallState == TRANSIENT_FAILURE) {
+            cancelFailOverTimer();
+            failOver();
```

Comment:
This two lines together is confusing. The only case you need to cancel the old failover and restart a new failover with current priority is when recomputing `currentPriority` from `updateLocalities(...)`. So why not do 

```java
cancelFailoverTimer();
currentPriority = -1;
failOver();
```

as I mentioned before?

Reply:
>The only case you need to cancel the old failover and restart a new failover with current priority is when recomputing currentPriority from updateLocalities(...).

This is not true. In the spec:

> ... the xds policy will start a timer for 10 seconds to allow the new locality map time to establish connections to its endpoints.  If the locality map has not reported state READY by the time the timer fires, or if the locality map reports state TRANSIENT_FAILURE before the timer fires, the xds policy will consider the map unreachable and will switch to a locality map for the priority N+1 localities.

This is exactly following the spec, why is it confusing? On the contrary, 

```
cancelFailoverTimer();
currentPriority = -1;
failOver();
```
Is a weird behavior/optimization that is not specified, and behaves differently from the spec on edge cases.

Reply:
Oh you are right. You would need to cancel timer when getting TRANSIENT_FAILURE.